ring/batch.go

```diff
@@ -25,23 +56,25 @@ type instance struct {
 }
 
 type itemTracker struct {
-	minSuccess  int
-	maxFailures int
-	succeeded   atomic.Int32
-	failed4xx   atomic.Int32
-	failed5xx   atomic.Int32
-	remaining   atomic.Int32
-	err         atomic.Error
+	minSuccess   int
+	maxFailures  int
+	succeeded    atomic.Int32
+	failedClient atomic.Int32
+	failedServer atomic.Int32
+	remaining    atomic.Int32
+	err          atomic.Error
+	done         atomic.Bool
 }
 
 func (i *itemTracker) recordError(err error) int32 {
 	i.err.Store(err)
 
-	if s, ok := status.FromError(err); ok && s.Code()/100 == 4 {
-		return i.failed4xx.Inc()
+	e := newError(err)
+	if e.IsClientError() {
```

Comment:
if we throw `e` away immediately, can we just have one function which returns a boolean? would reduce garbage